### PR TITLE
Support generated coupons in email editor

### PR DESCRIPTION
**Pull request by pavel-mailpoet**
## Description

Adds MailPoet support for WooCommerce’s generated Coupon Code block in the block email editor.

This supports two send paths:

1. Automation emails sent to one subscriber at a time can generate a per-recipient coupon and optionally restrict that coupon to the recipient’s email address.
2. Regular newsletters can generate a shared coupon for the send. Recipient-restricted generated coupons stay blocked for newsletters because there is no single recipient email for a bulk render.

MailPoet validates unsupported sends before rendering and stops sending when coupon generation fails, so users get a clear failure instead of an email with a broken coupon.

## Code review notes

This depends on the WooCommerce Coupon Code block auto-generation contract from https://github.com/woocommerce/woocommerce/pull/64342, currently targeted for WooCommerce 10.8.0. Users do not need to update WooCommerce and MailPoet at the same time: on older WooCommerce versions the generated-coupon UI/contract is not available, generated-coupon patterns are hidden, and MailPoet keeps the saved-content/rendering path defensive.

WooCommerce can register the Coupon Code block before or after MailPoet’s email editor integration runs. The JS extension handles both paths: it registers the normal block-type filter and also applies a bounded retry for an already-registered block so `restrictToSubscriber` serializes reliably.

The discount email patterns now explicitly save `source: "createNew"` and are only registered for WooCommerce 10.8.0 or newer. That avoids exposing generated-coupon-dependent pattern content to users whose WooCommerce version only has the older static coupon block contract.

MailPoet also patches the coupon block email render callback during email rendering when WooCommerce exposes the generated coupon block attributes but has not wired the matching email-editor coupon renderer. This keeps real sends working with WooCommerce 10.8 development builds where the UI is available but the block would otherwise keep `XXXX-XXXXXX-XXXX` in the delivered email.

### Prerequisites

- Use a MailPoet local environment with WooCommerce active.
- To test the generated coupon flow, use a WooCommerce build that includes https://github.com/woocommerce/woocommerce/pull/64342, expected in WooCommerce 10.8.0.
- Mailpit should be running so delivered emails can be inspected at http://localhost:8082/.
- Before each send test, it is helpful to clear Mailpit and note the current WooCommerce coupon count under WooCommerce → Marketing → Coupons.

### Regular newsletter: shared generated coupon

1. Go to MailPoet → Emails and create a regular newsletter.
2. Open it in the block email editor and pick any template.
3. Insert a WooCommerce Coupon Code block.
4. Leave the default values: Coupon source = Create new, percentage discount, amount 10, expiry 10 days.
5. Save the email and send it to a list with more than one subscriber.
6. Open the delivered email in Mailpit.
7. Confirm the email shows a real generated coupon code instead of `XXXX-XXXXXX-XXXX`.
8. Confirm WooCommerce created exactly one new coupon for that send.
9. Confirm the generated coupon is not restricted to a specific recipient email address.

Expected result: all recipients of that newsletter receive the same generated coupon code, and that code exists in WooCommerce.

### Older WooCommerce compatibility

1. Test with WooCommerce older than 10.8.0.
2. Confirm generated-coupon discount patterns are not registered.
3. Confirm older static/existing Coupon Code block behaviour is unchanged.
4. Confirm MailPoet does not require WooCommerce and MailPoet to be updated at the same time.

Expected result: generated coupon behaviour is unavailable on older WooCommerce versions, but existing coupon flows continue working.

## Linked PRs

- https://github.com/woocommerce/woocommerce/pull/64342

## Linked tickets

- https://linear.app/a8c/issue/STOMAIL-7969/coupon-auto-generation-block-for-the-gutenberg-email-editor

## After-merge notes

The generated-coupon editor flow requires WooCommerce 10.8.0 or later, assuming WooCommerce PR 64342 ships in that milestone. MailPoet remains compatible with older WooCommerce versions, but those users will not get the generated Coupon Code block behavior until WooCommerce is updated.

After WooCommerce 10.8.0 is available, do one final smoke test with the released WooCommerce build. MailPoet’s render fallback should no-op when WooCommerce wires its own email renderer correctly, and it should only matter for builds where the generated coupon UI exists but the email renderer is not connected.

### The editor

Screenshot 2026-04-28 at 10 32 33

### Rendered email

Screenshot 2026-04-28 at 11 17 07

## Tasks

- [x] I have added a changelog entry (`pnpm changelog:add --type=<type> --description=<description>`)
- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes

## Preview

[Preview in WordPress Playground](https://account.mailpoet.com/playground/new/branch:add/STOMAIL-7969-gutenberg-coupon-generation)

_The latest successful build from `add/STOMAIL-7969-gutenberg-coupon-generation` will be used. If none is available, the link won't work._

No issues found.

### CODE REVIEW

File: mailpoet/assets/js/src/mailpoet-email-editor-integration/coupon-code-restrict-to-subscriber-control.tsx [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        <Fragment>
+          <BlockEdit {...props} />
+          <InspectorControls>
+            <PanelBody title={__('Usage restriction', 'mailpoet')}>
+              <ToggleControl
+                checked={Boolean(
+                  props.attributes[RESTRICT_TO_SUBSCRIBER_ATTRIBUTE],
+                )}
+                label={restrictToSubscriberCopy}
+                help={restrictToSubscriberCopy}
```
[the remark this answers is not in the document]
Comment (pavel-mailpoet) [reply mailpoet/assets/js/src/mailpoet-email-editor-integration/coupon-code-restrict-to-subscriber-control.tsx]: Fixed in 1b7b1b6fb3: label and help now use separate strings; help explains that the generated coupon can only be redeemed by the recipient email address.

File: mailpoet/lib/EmailEditor/Integrations/MailPoet/EditorPageRenderer.php [hunk 2]
```diff
@@ -196,6 +196,7 @@ public function render() {
       'mailpoet_installed_days_ago' => (int)$installedAtDiff->format('%a'),
       'mailpoet_is_automation_newsletter' => $isAutomationNewsletter,
       'mailpoet_automation_id' => $automationId,
+      'mailpoet_gutenberg_coupon_generation_available' => true,
```
[the remark this answers is not in the document]
Comment (pavel-mailpoet) [reply mailpoet/lib/EmailEditor/Integrations/MailPoet/EditorPageRenderer.php]: Resolved in 58226cb733 and 1b7b1b6fb3: the hardcoded availability flag was removed. The JS now detects WooCommerce support from the coupon block contract by checking for the source attribute, so there is no MailPoet feature flag or WooCommerce version gate here.

File: mailpoet/lib/EmailEditor/Integrations/MailPoet/Coupons/CouponBlockValidator.php [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    $minimumAmount = $this->validateOptionalNumber($attrs['minimumAmount'] ?? null, 'minimumAmount');
+    $maximumAmount = $this->validateOptionalNumber($attrs['maximumAmount'] ?? null, 'maximumAmount');
+    if ($minimumAmount !== '' && $maximumAmount !== '' && $maximumAmount < $minimumAmount) {
+      throw new GutenbergCouponValidationException('Maximum amount must be greater than or equal to minimum amount.');
+    }
+
+    $emailRestrictions = $this->validateEmailRestrictions($attrs['emailRestrictions'] ?? []);
+    if ($this->toBoolean($attrs['restrictToSubscriber'] ?? false)) {
+      $emailRestrictions[] = $recipientEmail;
+    }
```
[the remark this answers is not in the document]
Comment (pavel-mailpoet) [reply mailpoet/lib/EmailEditor/Integrations/MailPoet/Coupons/CouponBlockValidator.php:43]: Fixed in 1b7b1b6fb3: restrictToSubscriber now passes the recipient email through validateEmailRestrictions before appending, so it is sanitized and lowercased consistently before array_unique.

File: mailpoet/assets/js/src/mailpoet-email-editor-integration/index.ts [hunk 4]
```diff
@@ -144,6 +125,48 @@ addFilter(
...
+
+    registerPlugin('mailpoet-automation-save-button', {
+      render: AutomationSaveButton,
+      scope: 'woocommerce-email-editor',
+    });
+  }
+
+  initStripPostStatusOnSaveMiddleware();
+  initializeEditor('mailpoet-email-editor');
+  ensureRestrictToSubscriberAttributeRegisteredWhenAvailable();
```
[the remark this answers is not in the document]
Comment (pavel-mailpoet) [reply mailpoet/assets/js/src/mailpoet-email-editor-integration/index.ts]: Fixed in 1b7b1b6fb3: removed the post-initializeEditor retry call from index.ts. The extension registers before editor initialization, and the retry fallback remains in the extension registration path only.

File: mailpoet/lib/Cron/Workers/SendingQueue/Tasks/Newsletter.php [hunk 5]
```diff
@@ -263,6 +271,66 @@ public function preProcessNewsletter(NewsletterEntity $newsletter, ScheduledTask
...
+      [
+        'newsletter_id' => $newsletter->getId(),
+        'queue_id' => $queue->getId(),
+      ]
+    );
+    $this->newslettersRepository->setAsCorrupt($newsletter);
+    $this->sendingQueuesRepository->pause($queue);
+
+    throw NewsletterProcessingException::create()->withMessage($message);
+  }
```
[the remark this answers is not in the document]
Comment (pavel-mailpoet) [reply mailpoet/lib/Cron/Workers/SendingQueue/Tasks/Newsletter.php:321]: Fixed in 1b7b1b6fb3: stopGutenbergCouponPreProcessing now only performs side effects. renderNewsletterOrStop rethrows a NewsletterProcessingException with the original exception chained as previous.

**Review by lysyjan — APPROVED** [review]
Looks good to merge.

A few things that we could consider to solve, but it's non-blocking:

- No early `isWooCommerceActive()` gate in `Newsletter::preflightCouponBlockGeneration` or `CouponBlockGenerator::init()`. The runtime check inside `generate()` does catch a deactivated-WC scenario, but surfaces a misleading error. Adding the check upstream gives a clearer message.
- Test gaps worth filling — `CouponBlockFailureTranslator` and `CouponBlockAttributeParser` have no dedicated test files; `EmailContextBuilderTest` doesn't cover the standard-newsletter real-send branch (the production path for non-automation sends).

**Comment by pavel-mailpoet** [comment]
Thanks, Jan. I checked these and agree they’re non-blocking.

WooCommerce inactive is already handled safely during generation with a clear send failure, so I’d avoid adding more preflight plumbing for that edge case now. The translator/parser tests would be nice to have, but their behaviour is covered through the generator/renderer paths.
